client.go

```diff
@@ -497,128 +523,92 @@ func (c *TunaSessionClient) DialWithConfig(remoteAddr string, config *nkn.DialCo
-		defer cancel()
-	}
-
-	buf, err := json.Marshal(&Request{Action: "getPubAddr"})
-	if err != nil {
-		return nil, err
-	}
-
-	respChan, err := c.multiClient.Send(nkn.NewStringArray(remoteAddr), buf, nil)
-	if err != nil {
-		return nil, err
-	}
-
-	var msg *nkn.Message
-	select {
-	case msg = <-respChan.C:
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	}
-
-	pubAddrs := &PubAddrs{}
-	err = json.Unmarshal(msg.Data, pubAddrs)
-	if err != nil {
-		return nil, err
+	var pubAddrs *PubAddrs
+	for { // try until listeners tuna exits are all ready.
+		pubAddrs, err = c.RequestPubAddrs(remoteAddr)
+		if err != nil {
+			return nil, err
+		}
+		if pubAddrs == nil || pubAddrs.Addrs == nil || len(pubAddrs.Addrs) == 0 {
+			time.Sleep(2 * time.Second)
+		} else {
+			break
+		}
 	}
 
 	sessionID, err := nkn.RandomBytes(SessionIDSize)
 	if err != nil {
 		return nil, err
 	}
 
-	var lock sync.Mutex
 	var wg sync.WaitGroup
-	conns := make(map[string]*Conn, len(pubAddrs.Addrs))
-	dialer := &net.Dialer{}
 	for i := range pubAddrs.Addrs {
 		wg.Add(1)
 		go func(i int) {
 			defer wg.Done()
-
-			if pubAddrs.Addrs[i] == nil || len(pubAddrs.Addrs[i].IP) == 0 || pubAddrs.Addrs[i].Port == 0 {
-				return
-			}
-
-			netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port))
-			if err != nil {
-				log.Printf("Dial error: %v", err)
-				return
-			}
-
-			conn := newConn(netConn)
-
-			err = writeMessage(conn, []byte(c.addr.String()), time.Duration(config.DialTimeout)*time.Millisecond)
+			_, err := c.dialAConn(sessionID, i, pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port)
 			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
+				log.Printf("tuna session, dialAConn err: %v\n", err)
 				return
 			}
-
-			metadata := &pb.SessionMetadata{
-				Id: sessionID,
-			}
-			metadataRaw, err := proto.Marshal(metadata)
-			if err != nil {
-				log.Printf("Encode session metadata error: %v", err)
-				conn.Close()
-				return
-			}
-
-			buf, err := c.encode(metadataRaw, remoteAddr)
-			if err != nil {
-				log.Printf("Encode message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			err = writeMessage(conn, buf, time.Duration(config.DialTimeout)*time.Millisecond)
-			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			lock.Lock()
-			conns[connID(i)] = conn
-			lock.Unlock()
 		}(i)
 	}
 	wg.Wait()
 
+	sessKey := sessionKey(remoteAddr, sessionID)
+	c.RLock()
+	conns := c.sessionConns[sessKey]
+	c.RUnlock()
 	connIDs := make([]string, 0, len(conns))
 	for i := 0; i < len(pubAddrs.Addrs); i++ {
 		if _, ok := conns[connID(i)]; ok {
 			connIDs = append(connIDs, connID(i))
 		}
 	}
 
-	sessKey := sessionKey(remoteAddr, sessionID)
 	sess, err := c.newSession(remoteAddr, sessionID, connIDs, config.SessionConfig)
 	if err != nil {
 		return nil, err
 	}
 
 	c.Lock()
 	c.sessions[sessKey] = sess
-	c.sessionConns[sessKey] = conns
```

Comment:
c.sessions and c.sessionConns need to be synchronized

Reply:
Synchronized means that, they need to be modified within the same lock period. Otherwise there might be a moment they are not consistent